docs/appendix-A.rst

```diff
@@ -0,0 +1,1783 @@
+
+YAML:
+
+.. code-block:: yaml
+
+    - decl: const char * getCharPtr3()
+      format:
+        F_string_result_as_arg: output
+
+The C wrapper:
+
+.. literalinclude:: ../regression/reference/strings/wrapstrings.cpp
+   :language: c
+   :start-after: start STR_get_char_ptr3_bufferify
+   :end-before: end STR_get_char_ptr3_bufferify
+
+Calls C via the interface:
+
+.. literalinclude:: ../regression/reference/strings/wrapfstrings.f
+   :language: fortran
+   :start-after: start c_get_char_ptr3_bufferify
+   :end-before: end c_get_char_ptr3_bufferify
+   :dedent: 8
+
+The Fortran wrapper:
+
+.. literalinclude:: ../regression/reference/strings/wrapfstrings.f
+   :language: fortran
+   :start-after: start get_char_ptr3
+   :end-before: end get_char_ptr3
+   :dedent: 4
+
+Fortran usage:
+
+.. code-block:: fortran
+
+    character(30) str
+    call get_char_ptrs(str)
+
+string functions
+----------------
+
+.. ############################################################
+
+.. _example_getConstStringRefPure:
+
+getConstStringRefPure
+^^^^^^^^^^^^^^^^^^^^^
+
+.. literalinclude:: ../regression/run/strings/strings.cpp
+   :language: c
+   :start-after: start getConstStringRefPure
+   :end-before: end getConstStringRefPure
+
+YAML:
+
+.. code-block:: yaml
+
+    - decl: const string& getConstStringRefPure()
+
+The C wrapper:
+
+.. literalinclude:: ../regression/reference/strings/wrapstrings.cpp
+   :language: c
+   :start-after: start STR_get_const_string_ref_pure_bufferify
+   :end-before: end STR_get_const_string_ref_pure_bufferify
+
+The native C wrapper:
+
+.. literalinclude:: ../regression/reference/strings/wrapstrings.cpp
+   :language: c
+   :start-after: start STR_get_const_string_ref_pure
+   :end-before: end STR_get_const_string_ref_pure
+
+Calls C via the interface:
+
+.. literalinclude:: ../regression/reference/strings/wrapfstrings.f
+   :language: fortran
+   :start-after: start c_get_const_string_ref_pure_bufferify
+   :end-before: end c_get_const_string_ref_pure_bufferify
+   :dedent: 8
+
+The Fortran wrapper:
+
+.. literalinclude:: ../regression/reference/strings/wrapfstrings.f
+   :language: fortran
+   :start-after: start get_const_string_ref_pure
+   :end-before: end get_const_string_ref_pure
+   :dedent: 4
+
+Fortran usage:
+
+.. code-block:: fortran
+
+    str = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
+    str = get_const_string_ref_pure()
+    call assert_true( str == static_str, "getConstStringRefPure")
+
+
+std::vector
+-----------
+
+.. ############################################################
+
+.. _example_vector_sum:
+
+vector_sum
+^^^^^^^^^^
+
+.. literalinclude:: ../regression/run/vectors/vectors.cpp
+   :language: c
+   :start-after: start vector_sum
+   :end-before: end vector_sum
+
+YAML:
+
+.. code-block:: yaml
+
+    - decl: int vector_sum(const std::vector<int> &arg)
+
+``intent(in)`` is implied for the *vector_sum* argument since it is
+``const``.  The Fortran wrapper passes the array and the size to C.
+
+The C wrapper:
```

Comment:
Does this always imply a copy of the vector's data?

(I guess there is no alternative since vector always owns its data).

```cpp
int VEC_vector_sum_bufferify(const int * arg, long Sarg)
{
// splicer begin function.vector_sum_bufferify
    const std::vector<int> SH_arg(arg, arg + Sarg);
    int SHC_rv = vector_sum(SH_arg);
    return SHC_rv;
// splicer end function.vector_sum_bufferify
}
```

Reply:
I added more detail and code samples to discuss memory management.  The examples I give will always copy data.  It was intended for a specific case that Dave Stevens has that uses very short vectors.  I plan to add an example that allows the std::vector to survive past the Fortran call, but it requires more work on the user's part. The user would get a Fortran pointer into the std::vector, but then have to responsible to manually free the memory. A work-in-progress.

When passing in an array, I looked at using a std::vector with an allocator that gives it the fortran memory. The trouble is, if you push_back beyond the Fortran memory you'll get an error when it tries to reallocate.  It may be useful in some situations since it can avoid copying.  But I've not pursued it further.